cpp/include/raft/error.hpp

```diff
@@ -0,0 +1,329 @@
+    : std::runtime_error(message) {}
+};
+
+/**
+ * @brief Exception thrown when a NCCL error is encountered.
+ */
+struct nccl_error : public std::runtime_error {
+  explicit nccl_error(char const* const message)
+    : std::runtime_error(message) {}
+  explicit nccl_error(std::string const& message)
+    : std::runtime_error(message) {}
+};
+
+}  // namespace raft
+
+#define STRINGIFY_DETAIL(x) #x
+#define RAFT_STRINGIFY(x) STRINGIFY_DETAIL(x)
+
+/**
+ * @brief Macro for checking (pre-)conditions that throws an exception when
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define RAFT_EXPECTS(cond, reason)                         \
+  (!!(cond))                                               \
+    ? static_cast<void>(0)                                 \
+    : throw raft::logic_error("RAFT failure at: " __FILE__ \
+                              ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Indicates that an erroneous code path has been taken.
+ *
+ * In host code, throws a `raft::logic_error`.
+ *
+ * @param[in] reason String literal description of the reason
+ */
+#define RAFT_FAIL(reason)                              \
+  throw raft::logic_error("RAFT failure at: " __FILE__ \
+                          ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Macro for checking (pre-)conditions that throws an exception when
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define CUML_EXPECTS(cond, reason)                         \
+  (!!(cond))                                               \
+    ? static_cast<void>(0)                                 \
+    : throw raft::logic_error("cuML failure at: " __FILE__ \
+                              ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Indicates that an erroneous code path has been taken.
+ *
+ * In host code, throws a `raft::logic_error`.
+ *
+ * @param[in] reason String literal description of the reason
+ */
+#define CUML_FAIL(reason)                              \
+  throw raft::logic_error("cuML failure at: " __FILE__ \
+                          ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Macro for checking (pre-)conditions that throws an exception when
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define CUGRAPH_EXPECTS(cond, reason)                         \
```

Comment:
Do we really need spearate RAFT/CUGRAPH/CUML EXPECT (or TRY), it seems they are all doing the same thing. Do you see a divergence in the future? I'd be fine using RAFT_EXPECTS in cuGraph.

Reply:
RAFT_EXPECTS prints a message starting with "RAFT failure at"
CUML(GRAPH)_EXPECTS prints a message starting with "CUML(CUGRAPH) failure at"

So these are not identical and I think this information can be helpful, what do you think?

Reply:
I would also prefer to keep CUML/CUGRAPH-specifics out of raft as much as possible. If this repository is successful and ends up feeding many different libraries in the RAPIDS ecosystem, I am concerned this approach will become unsustainable. 

What's the intended use for these? For example, will `CUGRAPH_EXPECTS` be getting used in prims like spectral clustering while `CUML_EXPECTS` will get used in dense linear algebra prims?

Reply:
I too agree with Alex and Corey on this one. Anything specific to cugraph or cuml is better put inside the respective library itself.

Reply:
Got it :-)